File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Pair{
    public:
    int dia;
    int ht;
};
class Solution {
public:
    Pair* diameter(TreeNode* root){
        if(root == NULL){
            Pair* bp = new Pair();
            bp->ht = -1;
            bp->dia = 0;
            return bp;
        }

        Pair* left = diameter(root->left);
        Pair* right = diameter(root->right);

        Pair* mp = new Pair();
        mp->ht = max(left->ht, right->ht) + 1;

        int candidate = left->ht + right->ht + 2;
        mp->dia = max({candidate, left->dia, right->dia});
        return mp;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        Pair* mp = diameter(root);
        return mp->dia;
    }
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class DiaPair{
    int diameter;
    int height;
}

class Solution {
    public DiaPair Diameter(TreeNode root){
        if(root == null){
            DiaPair bp = new DiaPair();
            bp.diameter = 0;
            bp.height = -1;
            return bp;
        }

        DiaPair left = Diameter(root.left);
        DiaPair right = Diameter(root.right);

        DiaPair mp = new DiaPair();
        mp.height = Math.max(left.height, right.height) + 1;
        int freq = left.height + right.height + 2;

        mp.diameter = Math.max(freq, Math.max(left.diameter, right.diameter));
        return mp;
    }
    public int diameterOfBinaryTree(TreeNode root) {
        DiaPair mp = Diameter(root);
        return mp.diameter;
    }
}

